File: java/src/main/java/com/danifoldi/dml/DmlParser.java
package com.danifoldi.dml;

import com.danifoldi.dml.type.DmlArray;
import com.danifoldi.dml.type.DmlBoolean;
import com.danifoldi.dml.type.DmlComment;
import com.danifoldi.dml.type.DmlKey;
import com.danifoldi.dml.type.DmlNumber;
import com.danifoldi.dml.type.DmlObject;
import com.danifoldi.dml.type.DmlDocument;
import com.danifoldi.dml.type.DmlString;
import com.danifoldi.dml.type.DmlValue;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.math.BigDecimal;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.stream.Collectors;

public class DmlParser {
    private final String dmlString;
    private int pointer;
    private int line;
    private int column;

    private DmlParser(String dmlString) {
        this.dmlString = dmlString + "\n";
        line = 1;
        column = 1;
        pointer = 0;
    }

    private void step() {
        if (currentChar() == '\n') {
            column = 0;
            line++;
        }

        column++;
        pointer++;
    }

    private void step(int count) {
        for (int i = 0; i < count; i++) {
            step();
        }
    }

    private void stepNextNonWhitespace() {
        while (pointer < dmlString.length() && isWhitespace(currentChar())) {
            step();
        }
    }

    private int nextNonWhitespace(int pointer) {
        while (pointer < dmlString.length() && isWhitespace(charAt(pointer))) {
            pointer++;
        }

        return pointer;
    }

    private char nextNonWhitespaceChar(int pointer) {
        while (pointer < dmlString.length() && isWhitespace(charAt(pointer))) {
            pointer++;
        }

        return charAt(pointer);
    }

    private char charAt(int pointer) {
        return dmlString.charAt(pointer);
    }

    private char currentChar() {
        return charAt(pointer);
    }

    private char nextChar() {
        return charAt(pointer + 1);
    }

    private boolean isCommentStart(char c, char n) {
        return c == '/' && (n == '/' || n == '*');
    }

    private boolean isMultilineCommentEnd(char c, char n) {
        return c == '*' && n == '/';
    }

    private boolean isWhitespace(char c) {
        return String.valueOf(c).matches("\\s");
    }

    private boolean isDigit(char c) {
        return String.valueOf(c).matches("[0-9]");
    }

    private boolean isNewline(char c) {
        return c == '\n';
    }

    private boolean isString(char c) {
        return c == '"' || c == '\'';
    }

    private boolean isKey(char c) {
        return String.valueOf(c).matches("[a-zA-Z0-9-_]");
    }


    private DmlComment parseComment() throws DmlParseException {
        StringBuilder comment = new StringBuilder();

        stepNextNonWhitespace();

        while (isCommentStart(currentChar(), nextChar())) {
            switch (nextChar()) {
                case '/' -> {
                    step();
                    step();
                    int start = pointer;
                    while (!isNewline(currentChar())) {
                        step();
                    }
                    comment.append(dmlString, start, pointer);
                }
                case '*' -> {
                    step();
                    step();
                    int start = pointer;
                    while (!isMultilineCommentEnd(currentChar(), nextChar())) {
                        step();
                    }
                    comment.append(dmlString, start, pointer);
                    step();
                    step();
                }
            }
        }

        return new DmlComment(comment.toString());
    }

    private DmlBoolean parseBoolean(DmlComment comment) throws DmlParseException {
        if (dmlString.substring(pointer, pointer + 4).equalsIgnoreCase("true")) {
            step(4);
            DmlBoolean bool = new DmlBoolean(true);
            bool.comment(comment);
            return bool;
        }
        if (dmlString.substring(pointer, pointer + 5).equalsIgnoreCase("false")) {
            step(5);
            DmlBoolean bool = new DmlBoolean(false);
            bool.comment(comment);
            return bool;
        }
        throw new DmlParseException(line, column, "true or false", dmlString.substring(pointer, pointer + 4));
    }

    private DmlNumber parseNumber(DmlComment comment) throws DmlParseException {
        stepNextNonWhitespace();
        int n = pointer;
        boolean negative = currentChar() == '-';
        if (negative) {
            step();
        }
        String whole;
        String fraction = "0";
        String exponent = "0";

        int start = pointer;
        while (isDigit(currentChar())) {
            step();
        }
        whole = pointer > start ? dmlString.substring(start, pointer) : "0";
        if (currentChar() == '.') {
            step();
            start = pointer;
            while (isDigit(currentChar())) {
                step();
            }
            fraction = pointer > start ? dmlString.substring(start, pointer) : "0";
        }
        if (currentChar() == 'e') {
            step();
            start = pointer;
            while (isDigit(currentChar())) {
                step();
            }
            boolean nexp = currentChar() == '-';
            if (nexp) {
                step();
            }
            exponent = (nexp ? "-" : "") + (pointer > start ? dmlString.substring(start, pointer) : "0");
        }
        BigDecimal b = new BigDecimal(whole + "." + fraction);
        if (negative) {
            b = b.negate();
        }
        b = b.scaleByPowerOfTen(Integer.parseInt(exponent));
        DmlNumber number = new DmlNumber(b, dmlString.substring(n, pointer));
        number.comment(comment);
        return number;
    }

    private DmlString parseString(DmlComment comment) throws DmlParseException {
        stepNextNonWhitespace();
        switch (charAt(pointer)) {
            case '\'' -> {
                int start = pointer;
                boolean didEscape;
                do {
                    didEscape = currentChar() == '\\';
                    step();
                } while (currentChar() != '\'' || didEscape);
                step();
                DmlString string = new DmlString(dmlString.substring(start + 1, pointer - 1).replace("\\'", "'"));
                string.comment(comment);
                return string;
            }
            case '"' -> {
                int start = pointer;
                boolean didEscape;
                do {
                    didEscape = currentChar() == '\\';
                    step();
                } while (currentChar() != '"' || didEscape);
                step();
                DmlString string = new DmlString(dmlString.substring(start + 1, pointer - 1).replace("\\\"", "\""));
                string.comment(comment);
                return string;
            }
            default -> throw new DmlParseException(line, column, "' or \"", String.valueOf(charAt(pointer)));
        }
    }

    private DmlKey parseKey(DmlComment comment) throws DmlParseException {
        stepNextNonWhitespace();
        int start = pointer;
        while (isKey(charAt(pointer))) {
            step();
        }
        DmlKey key = new DmlKey(dmlString.substring(start, pointer));
        key.comment(comment);
        return key;
    }

    private DmlValue parseValue(DmlComment comment) throws DmlParseException {
        stepNextNonWhitespace();
        return switch (charAt(pointer)) {
            case '{' -> parseObject(comment);
            case '[' -> parseArray(comment);
            case '\'', '"' -> parseString(comment);
            case 't', 'f' -> parseBoolean(comment);
            default -> parseNumber(comment);
        };
    }

    private DmlObject parseObjectValue(boolean toplevel) throws DmlParseException {
        DmlObject object = new DmlObject(new HashMap<>());
        while (charAt(nextNonWhitespace(pointer)) != '}' && pointer < dmlString.length()) {
            stepNextNonWhitespace();
            DmlComment comment = parseComment();
            stepNextNonWhitespace();
            DmlKey key = parseKey(comment);
            stepNextNonWhitespace();
            if (currentChar() != ':') {
                throw new DmlParseException(line, column, ":", String.valueOf(currentChar()));
            }
            step();
            stepNextNonWhitespace();
            DmlComment comment1 = parseComment();
            stepNextNonWhitespace();
            DmlValue value = parseValue(comment1);
            object.set(key, value);
            int sline = line;
            stepNextNonWhitespace();
            if ((pointer >= dmlString.length() && toplevel) || currentChar() == '}') {
                break;
            }
            if (currentChar() != ',' && sline == line) {
                throw new DmlParseException(line, column, ",", String.valueOf(currentChar()));
            }
            if (currentChar() == ',') {
                step();
            }
        }
        return object;
    }

    private DmlObject parseObject(DmlComment comment) throws DmlParseException {
        step();
        if (comment == null) {
            comment = parseComment();
        }
        DmlObject object = parseObjectValue(false);
        object.comment(comment);

        if (charAt(nextNonWhitespace(pointer)) != '}') {
            throw new DmlParseException(line, column, "}", String.valueOf(charAt(nextNonWhitespace(pointer))));
        }
        stepNextNonWhitespace();
        step();
        return object;
    }

    private DmlArray parseArray(DmlComment comment) throws DmlParseException {
        step();
        if (comment == null) {
            comment = parseComment();
        }
        DmlArray array = new DmlArray(new ArrayList<>());
        array.comment(comment);

        while (charAt(nextNonWhitespace(pointer)) != ']') {
            stepNextNonWhitespace();
            DmlComment vcomment = parseComment();
            array.add(parseValue(vcomment));
            int sline = line;
            stepNextNonWhitespace();
            if (currentChar() == ']') {
                break;
            }
            if (currentChar() != ',' && sline == line) {
                throw new DmlParseException(line, column, ",", String.valueOf(currentChar()));
            }
            if (currentChar() == ',') {
                step();
            }
        }

        if (charAt(nextNonWhitespace(pointer)) != ']') {
            throw new DmlParseException(line, column, "]", String.valueOf(charAt(nextNonWhitespace(pointer))));
        }
        stepNextNonWhitespace();
        step();
        return array;
    }

    private DmlValue parseDocument() throws DmlParseException {
        stepNextNonWhitespace();
        DmlComment comment = parseComment();
        stepNextNonWhitespace();
        DmlValue result;

        switch (nextNonWhitespaceChar(pointer)) {
            case '{':
                result = parseObject(comment);
                break;
            case '[':
                result = parseArray(comment);
                break;
            default:
                if (isKey(charAt(pointer))) {
                    DmlObject object = parseObjectValue(true);
                    object.comment(comment);
                    result = object;
                } else {
                    throw new DmlParseException(line, column, "{, [, \", ' or key", String.valueOf(charAt(pointer)));
                }
        }

        if (nextNonWhitespace(pointer) < dmlString.length()) {
            throw new DmlParseException(line, column, "document end", String.valueOf(charAt(nextNonWhitespace(pointer))));
        }

        return result;
    }

    public static DmlValue parse(String dmlString) throws DmlParseException {
        return new DmlParser(dmlString).parseDocument();
    }

    public static DmlValue parse(Path file) throws DmlParseException, IOException {
        try (BufferedReader reader = Files.newBufferedReader(file)) {
            return parse(reader.lines().collect(Collectors.joining("\n")));
        }
    }

    public static String serialize(DmlDocument document) {
        return document.serializeDocument();
    }

    public static void serialize(DmlDocument document, Path file) throws IOException {
        try (BufferedWriter writer = Files.newBufferedWriter(file)) {
            writer.write(serialize(document));
        }
    }

    public static void main(String[] args) throws DmlParseException{
        String test = """
                // hello
                a: 1
                /*
                * asd
                */
                b: ["ran\\"d'om",1,2,3,4,5
                6,7,8
                9
                10
                
                ]
                /*
                anyád
                picsája +
                */
                c: {tomi: "van", egysenki: "há mé
                 lenne"}
                 e:[]
                 d:{}
                """;
        DmlObject o = (DmlObject)DmlParser.parse(test);
        System.out.println(o);
        System.out.println(DmlParser.serialize(o));
    }
}

File: javascript/src/DmlParser.js
import Big from 'big.js'

import DmlParseError from './error/DmlParseError.js'
import DmlArray from './types/DmlArray.js'
import DmlBoolean from './types/DmlBoolean.js'
import DmlComment from './types/DmlComment.js'
import DmlKey from './types/DmlKey.js'
import DmlNumber from './types/DmlNumber.js'
import DmlObject from './types/DmlObject.js'
import DmlString from './types/DmlString.js'

class DmlParser {
  #dmlString
  #line
  #column
  #pointer
  constructor(dmlString) {
    this.#dmlString = dmlString
    this.#line = 1
    this.#column = 1
    this.#pointer = 0
  }

  #step(count = 1) {
    for (let i = 0; i < count; i++) {
      if (this.#currentChar() == '\n') {
        this.#column = 0
        this.#line++
      }
      this.#column++
      this.#pointer++
    }
  }

  #stepNextNonWhitespace() {
    while (this.#pointer < this.#dmlString.length && this.#isWhitespace(this.#currentChar())) {
      this.#step()
    }
  }

  #nextNonWhitespace(pointer) {
    while (pointer < this.#dmlString.length && this.#isWhitespace(this.#charAt(pointer))) {
      pointer++
    }

    return pointer
  }

  #charAt(pointer) {
    return this.#dmlString.charAt(pointer)
  }

  #currentChar() {
    return this.#dmlString.charAt(this.#pointer)
  }

  #nextChar() {
    return this.#dmlString.charAt(this.#pointer + 1)
  }

  #isCommentStart(c, n) {
    return c === '/' && (n === '/' || n === '*')
  }

  #isMultilineCommentEnd(c, n) {
    return c === '*' && n === '/'
  }

  #isWhitespace(c) {
    return /\s/.test(c)
  }

  #isDigit(c) {
    return /[0-9]/.test(c)
  }

  #isNewline(c) {
    return c === '\n'
  }

  #isKey(c) {
    return /[a-zA-Z0-9-_+]/.test(c)
  }

  #parseComment() {
    let comment = ''
    this.#stepNextNonWhitespace()

    while (this.#isCommentStart(this.#currentChar(), this.#nextChar())) {
      switch (this.#nextChar()) {
        case '/': {
          this.#step(2)
          let start = this.#pointer
          while (!this.#isNewline(this.#currentChar())) {
            this.#step()
          }
          comment += this.#dmlString.substring(start, this.#pointer)
          break
        }
        case '*': {
          this.#step(2)
          let start = this.#pointer
          while (!this.#isMultilineCommentEnd(this.#currentChar(), this.#nextChar())) {
            this.#step()
          }
          comment += this.#dmlString.substring(start, this.#pointer)
          this.#step(2)
          break
        }
      }
    }

    return new DmlComment(comment)
  }

  #parseBoolean(comment) {
    if (this.#dmlString.substring(this.#pointer, this.#pointer + 4).toLowerCase() === 'true') {
      this.#step(4)
      return new DmlBoolean(true).withComment(comment)
    }
    if (this.#dmlString.substring(this.#pointer, this.#pointer + 5).toLowerCase() === 'false') {
      this.#step(5)
      return new DmlBoolean(false).withComment(comment)
    }
    throw new DmlParseError(
      this.#line,
      this.#column,
      'true or false',
      this.#dmlString.substring(this.#pointer, this.#pointer + 4)
    )
  }

  #parseNumber(comment) {
    this.#stepNextNonWhitespace()
    let n = this.#pointer
    let negative = this.#currentChar() === '-'
    if (negative) {
      this.#step()
    }
    let whole = ''
    let fraction = '0'
    let exponent = '0'

    let start = this.#pointer
    while (this.#isDigit(this.#currentChar())) {
      this.#step()
    }
    whole = this.#pointer > start ? this.#dmlString.substring(start, this.#pointer) : '0'
    if (this.#currentChar() == '.') {
      this.#step()
      start = this.#pointer
      while (this.#isDigit(this.#currentChar())) {
        this.#step()
      }
      fraction = this.#pointer > start ? this.#dmlString.substring(start, this.#pointer) : '0'
    }
    if (this.#currentChar() === 'e') {
      this.#step()
      start = this.#pointer
      while (this.#isDigit(this.#currentChar())) {
        this.#step()
      }
      let nexp = this.#currentChar() === '-'
      if (nexp) {
        this.#step()
      }
      exponent = (nexp ? '-' : '') + this.#pointer > start ? this.#dmlString.substring(start, this.#pointer) : '0'
    }
    let b = new Big(whole + '.' + fraction)
    if (negative) {
      b = b.times(-1)
    }
    b = b.pow(parseInt(exponent))
    return new DmlNumber(b, this.#dmlString.substring(n, this.#pointer)).withComment(comment)
  }

  #parseString(comment) {
    this.#stepNextNonWhitespace()
    switch (this.#currentChar()) {
      case '\'': {
        let start = this.#pointer
        let didEscape = false
        do {
          didEscape = this.#currentChar() === '\\'
          this.#step()
        } while (this.#currentChar() !== '\'' || didEscape)
        this.#step()
        return new DmlString(this.#dmlString.substring(start + 1, this.#pointer - 1).replace('\\\'', '\''))
          .withComment(comment)
      }
      case '"': {
        let start = this.#pointer
        let didEscape = false
        do {
          didEscape = this.#currentChar() === '\\'
          this.#step()
        } while (this.#currentChar() !== '"' || didEscape)
        this.#step()
        return new DmlString(this.#dmlString.substring(start + 1, this.#pointer - 1).replace('\\"', '"'))
          .withComment(comment)
      }
      default:
        throw new DmlParseError(this.#line, this.#column, '\' or "', this.#currentChar())
    }
  }

  #parseKey(comment) {
    this.#stepNextNonWhitespace()
    let start = this.#pointer
    while (this.#isKey(this.#currentChar())) {
      this.#step()
    }
    return new DmlKey(this.#dmlString.substring(start, this.#pointer)).withComment(comment)
  }

  #parseValue(comment) {
    this.#stepNextNonWhitespace()
    switch (this.#currentChar()) {
      case '{':
        return this.#parseObject(comment)
      case '[':
        return this.#parseArray(comment)
      case '\'':
      case '"':
        return this.#parseString(comment)
      case 't':
      case 'f':
        return this.#parseBoolean(comment)
      default:
        return this.#parseNumber(comment)
    }
  }

  #parseObjectValue(topLevel) {
    let object = new DmlObject(new Map())
    while (this.#charAt(this.#nextNonWhitespace(this.#pointer)) !== '}' && this.#pointer < this.#dmlString.length) {
      this.#stepNextNonWhitespace()
      let comment = this.#parseComment()
      this.#stepNextNonWhitespace()
      let key = this.#parseKey(comment)
      this.#stepNextNonWhitespace()
      if (this.#currentChar() !== ':') {
        throw new DmlParseError(this.#line, this.#column, ':', this.#currentChar())
      }
      this.#step()
      this.#stepNextNonWhitespace()
      let comment1 = this.#parseComment()
      this.#stepNextNonWhitespace()
      let value = this.#parseValue(comment1)
      object.set(key, value)
      let sline = this.#line
      this.#stepNextNonWhitespace()
      if ((this.#pointer >= this.#dmlString.length && topLevel) || this.#currentChar() === '}') {
        break
      }
      if (this.#currentChar() !== ',' && sline == this.#line) {
        throw new DmlParseError(this.#line, this.#column, ',', this.#currentChar())
      }
      if (this.#currentChar() === ',') {
        this.#step()
      }
    }
    return object
  }

  #parseObject(comment) {
    this.#step()
    if (comment == null) {
      comment = this.#parseComment()
    }
    let object = this.#parseObjectValue(false).withComment(comment)

    if (this.#charAt(this.#nextNonWhitespace(this.#pointer)) !== '}') {
      throw new DmlParseError(this.#line, this.#column, '}', this.#charAt(this.#nextNonWhitespace(this.#pointer)))
    }
    this.#stepNextNonWhitespace()
    this.#step()
    return object
  }

  #parseArray(comment) {
    this.#step()
    if (comment == null) {
      comment = this.#parseComment()
    }
    let array = new DmlArray([]).withComment(comment)

    while (this.#charAt(this.#nextNonWhitespace(this.#pointer)) !== ']') {
      this.#stepNextNonWhitespace()
      let comment1 = this.#parseComment()
      this.#stepNextNonWhitespace()
      let value = this.#parseValue(comment1)
      array.add(value)
      let sline = this.#line
      this.#stepNextNonWhitespace()
      if (this.#pointer >= this.#dmlString.length || this.#currentChar() === ']') {
        break
      }
      if (this.#currentChar() !== ',' && sline == this.#line) {
        throw new DmlParseError(this.#line, this.#column, ',', this.#currentChar())
      }
      if (this.#currentChar() === ',') {
        this.#step()
      }
    }

    if (this.#charAt(this.#nextNonWhitespace(this.#pointer)) !== ']') {
      throw new DmlParseError(this.#line, this.#column, ']', this.#charAt(this.#nextNonWhitespace(this.#pointer)))
    }
    this.#stepNextNonWhitespace()
    this.#step()
    return array
  }

  #parseDocument() {
    this.#stepNextNonWhitespace()
    let comment = this.#parseComment()
    this.#stepNextNonWhitespace()
    let result
    switch (this.#charAt(this.#stepNextNonWhitespace(this.#pointer))) {
      case '{':
        result = this.#parseObject(comment)
        break
      case '[':
        result = this.#parseArray(comment)
        break
      default:
        if (this.#isKey(this.#currentChar())) {
          result = this.#parseObjectValue(true).withComment(comment)
        } else {
          throw new DmlParseError(this.#line, this.#column, '{, [ or key', this.#currentChar())
        }
    }
    if (this.#nextNonWhitespace(this.#pointer) < this.#dmlString.length) {
      throw new DmlParseError(
        this.#line,
        this.#column,
        'document end',
        this.#charAt(this.#nextNonWhitespace(this.#pointer))
      )
    }

    return result
  }

  static parse(dmlString) {
    return new DmlParser(dmlString).#parseDocument()
  }
}

export default DmlParser
